### PR TITLE
Prevent forwarding and sync match when there is a backlog

### PR DESCRIPTION
**Pull request by ShahabT**
**What changed?**
- Added a config to define a threshold for *negligible* backlog based on the age of the oldest task in the backlog. The value is configurable per TQ, but planning to set it to 5 secs by default in prod. A backlog older than this threshold is considered non-negligible.
- Stop forwarding tasks when there is a non-negligible backlog
    - This in turn stops pollers to be forwarded in presence of a non-negligible backlog.
    - This ensures that all the partition accumulate and process backlog at the same rate. Effectively, the backlog age diff between partitions should not grow significantly larger than the negligibility threshold (e.g. 5 secs).
    - There is an exception to this: forwarding tasks is allowed when the partition has not received a poll recently (configurable, default 200ms). This is to ensure that in a TQ with few pollers, tasks and pollers are do not stuck in different partitions for a long time.
- Prevent sync match when there is a non-negligible backlog. This is already mostly happening (i.e. history tasks are all spooled and only synced matched in race conditions, when there is a backlog). However, forwarded tasks are also “sync match”ed in the root partition. In effect this change makes sure that the root partition does not dispatch forwarded tasks when itself has a backlog. This helps further with keeping the partitions balanced.
- Also added the following metrics to be able to monitor poll and task dispatch behavior at TQ partition level:
    - `poll_latency` with `forwarded=false/true`
    - `task_dispatch_latency` with `forwarded=false/true` and `source=History/DbBacklog`

**Why?**
With previous forwarding logic, when a backlog started to form, the root partition was effectively receiving more pollers and was able to process its backlog faster. This mixed up the order of dispatching tasks: a new task could be dispatched faster by going to the root partitions while old tasks are backlogged in other partitions. This change prevents that situation.

**How did you test it?**
- Unit tests
- Manually verified against 9 workload+worker configurations as defined here: https://github.com/ShahabT/temporal-tests/blob/main/balanced-partitions/experiments.go

**Potential risks**
The change is config protected and the default value is disabled (by setting a very high threshold). Will be enabled by default in a future release after successful verification in prod workloads.

**Is hotfix candidate?**
No.

### CODE REVIEW

File: common/metrics/metric_defs.go [hunk 1]
```diff
@@ -945,6 +945,7 @@ var (
 	PollSuccessPerTaskQueueCounter            = NewCounterDef("poll_success")
 	PollTimeoutPerTaskQueueCounter            = NewCounterDef("poll_timeouts")
 	PollSuccessWithSyncPerTaskQueueCounter    = NewCounterDef("poll_success_sync")
+	PollLatencyPerTaskQueue                   = NewCounterDef("poll_latency")
```
Comment (dnr) [inline common/metrics/metric_defs.go:948]: not NewTimerDef?

File: service/matching/matcher.go [hunk 2]
```diff
@@ -60,14 +62,18 @@ type TaskMatcher struct {
 	// rateLimiter that limits the rate at which tasks can be dispatched to consumers
 	rateLimiter quotas.RateLimiter
 
-	fwdr           *Forwarder
-	metricsHandler metrics.Handler // namespace metric scope
-	numPartitions  func() int      // number of task queue partitions
+	fwdr                   *Forwarder
+	metricsHandler         metrics.Handler // namespace metric scope
+	numPartitions          func() int      // number of task queue partitions
+	backlogTasksCreateTime map[int64]int
```
Comment (dnr) [inline service/matching/matcher.go:68]: comment what the key and value represents?

File: service/matching/matcher.go [hunk 3]
```diff
@@ -60,14 +62,18 @@ type TaskMatcher struct {
...
+	backlogTasksLock       sync.RWMutex
```
Comment (dnr) [inline service/matching/matcher.go:69]: my usual approach is to start with a Mutex and only use RWMutex if benchmarking shows a problem, or it's obviously very read-heavy. this seems read-write balanced so I'd just start with a Mutex

File: service/matching/matcher.go [hunk 4]
```diff
@@ -145,12 +152,25 @@ func (tm *TaskMatcher) Offer(ctx context.Context, task *internalTask) (bool, err
 		}
 	}
 
+	if !tm.isBacklogNegligible() {
```
Comment (dnr) [inline service/matching/matcher.go:155]: shouldn't we do this before we take a rate limiter token?

File: service/matching/matcher.go [hunk 5]
```diff
@@ -241,22 +262,58 @@ func (tm *TaskMatcher) MustOffer(ctx context.Context, task *internalTask, interr
...
 	select {
 	case tm.taskC <- task:
+		tm.emitDispatchLatency(task, false)
 		return nil
 	case <-ctx.Done():
 		return ctx.Err()
 	default:
 	}
 
+	reconsiderFwdTimer := (*time.Timer)(nil)
```
Comment (dnr) [inline service/matching/matcher.go:279]:
```suggestion
	var reconsiderFwdTimer *time.Timer
```

File: service/matching/matcher.go [hunk 6]
```diff
@@ -241,22 +262,58 @@ func (tm *TaskMatcher) MustOffer(ctx context.Context, task *internalTask, interr
...
+		if reconsiderFwdTimer != nil {
+			reconsiderFwdTimer.Stop()
+		}
+	}()
+
 forLoop:
 	for {
+		fwdTokenC := tm.fwdrAddReqTokenC()
+		reconsiderFwdTimer = (*time.Timer)(nil)
+		reconsiderFwdTimerC := (<-chan time.Time)(nil)
```
Comment (dnr) [inline service/matching/matcher.go:290]:
```suggestion
		reconsiderFwdTimer = nil
		var reconsiderFwdTimerC <-chan time.Time
```

File: service/matching/matcher.go [hunk 7]
```diff
@@ -286,26 +344,49 @@ forLoop:
...
+	source := task.source
+	if source == enumsspb.TASK_SOURCE_UNSPECIFIED {
+		// history may not specify the source
+		source = enumsspb.TASK_SOURCE_HISTORY
+	}
+
+	tm.metricsHandler.Timer(metrics.TaskDispatchLatencyPerTaskQueue.GetMetricName()).Record(
+		time.Since(*task.event.Data.CreateTime),
+		metrics.StringTag("source", source.String()),
+		metrics.StringTag("forwarded", strconv.FormatBool(forwarded)),
```
Comment (dnr) [inline service/matching/matcher.go:373]: I had a crazy idea on metrics here: what if we attached to each task and poll a short coded string. Polls start with the empty string, add "P" each time a poll is forwarded. Tasks start with "H" if direct from history or "D" if it passed through the db, and then add a "T" each time it's forwarded. When we get a match, concatenate the task and poll strings and use that as the metric label. It's one label that includes all the information we want concisely. We can use regexp matches to select different combinations.

File: service/matching/matcher.go [hunk 8]
```diff
@@ -409,22 +507,22 @@ func (tm *TaskMatcher) poll(ctx context.Context, pollMetadata *pollMetadata, que
...
 		tm.metricsHandler.Counter(metrics.PollSuccessWithSyncPerTaskQueueCounter.GetMetricName()).Record(1)
 		tm.metricsHandler.Counter(metrics.PollSuccessPerTaskQueueCounter.GetMetricName()).Record(1)
-		return task, nil
+		return task, false, nil
 	}
 }
 
 func (tm *TaskMatcher) fwdrPollReqTokenC() <-chan *ForwarderReqToken {
-	if tm.fwdr == nil {
+	if tm.fwdr == nil || !tm.isBacklogNegligible() {
```
Comment (dnr) [inline service/matching/matcher.go:525]: it feels a little weird to hide this here. I think I'd prefer if it was handed at the sites that call this, even if it's more verbose. in that case, maybe it makes sense to put the `tm.fwdr == nil` case there also and get rid of this?

File: service/matching/matcher.go [hunk 9]
```diff
@@ -440,3 +538,80 @@ func (tm *TaskMatcher) fwdrAddReqTokenC() <-chan *ForwarderReqToken {
 func (tm *TaskMatcher) isForwardingAllowed() bool {
 	return tm.fwdr != nil
 }
+
+// isBacklogNegligible returns true of the age of backlog is less than the threshold. Note that this relies on incoming
+// MustOffer calls being made back-to-back. If called between two subsequent MustOffer calls, it may return false positive.
```
Comment (dnr) [inline service/matching/matcher.go:543]: do you think we should add a comment to MustOffer explaining this restriction?

I did have the idea of offering a few tasks from the backlog at once to help with some of these issues. is that an idea that's still possibly helpful, or do we definitely not need it anymore?

File: service/matching/matcher.go [hunk 10]
```diff
@@ -60,14 +62,18 @@ type TaskMatcher struct {
...
 
-	fwdr           *Forwarder
-	metricsHandler metrics.Handler // namespace metric scope
-	numPartitions  func() int      // number of task queue partitions
+	fwdr                   *Forwarder
+	metricsHandler         metrics.Handler // namespace metric scope
+	numPartitions          func() int      // number of task queue partitions
+	backlogTasksCreateTime map[int64]int   // counting the number of tasks for each creation time
+	backlogTasksLock       sync.Mutex
+	lastPoller             atomic.Int64
```
Comment (dnr) [inline service/matching/matcher.go:70]: `// unix nanos of most recent poll start time`

**Review by dnr — APPROVED** [review]
approved, but a couple questions and comment suggestions

File: service/matching/matcher_test.go [hunk 11]
```diff
@@ -447,12 +447,13 @@ func (t *MatcherTestSuite) TestRemotePoll() {
 	}()
 
 	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
```
[the remark this answers is not in the document]
Comment (ShahabT) [reply service/matching/matcher_test.go:597]: added a bunch of unit tests

**Pull request merged**